psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,192 +768,882 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+
+            for (const auto& value : shell_values) {
+                sig_S_for_P[dP].push_back(std::get<0>(value));
+            }
+
+        }
+
+        // Significant R for Q
+        for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+            int dQ = Q - Qstart;
+            std::vector<std::tuple<int, double>> shell_values;
+            for (int R = Rstart; R < Rstart + nRshell; R++) {
+                int dR = R - Rstart;
+                double D_max = ints[0]->shell_pair_max_density(Q, R);
+                double screen_val = D_max * shell_ceiling_Q[dQ] * shell_ceiling_R[dR];
+                if (screen_val > linK_ints_cutoff_) {
+                    shell_values.push_back(std::make_tuple(R, screen_val));
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_R_for_Q[dQ].push_back(std::get<0>(value));
+            }
+
+        }
+
+        // Significant S for Q
+        for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+            int dQ = Q - Qstart;
+            std::vector<std::tuple<int, double>> shell_values;
+            for (int S = Sstart; S < Sstart + nSshell; S++) {
+                int dS = S - Sstart;
+                double D_max = ints[0]->shell_pair_max_density(Q, S);
+                double screen_val = D_max * shell_ceiling_Q[dQ] * shell_ceiling_S[dS];
+                if (screen_val > linK_ints_cutoff_) {
+                    shell_values.push_back(std::make_tuple(S, screen_val));
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_S_for_Q[dQ].push_back(std::get<0>(value));
+            }
+        }
+
+        // Stripeout Lists, used in the "Stripe out" section of the code
+        // Defined after the forthcoming loop
+        std::vector<std::unordered_set<int>> stripeout_R_for_P(nPshell);
+        std::vector<std::unordered_set<int>> stripeout_S_for_P(nPshell);
+        std::vector<std::unordered_set<int>> stripeout_R_for_Q(nQshell);
+        std::vector<std::unordered_set<int>> stripeout_S_for_Q(nQshell);
+
+        bool touched = false;
+        for (int P = Pstart; P < Pstart + nPshell; P++) {
+            for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+
+                if (Q > P) continue;
+
+                int dP = P - Pstart;
+                int dQ = Q - Qstart;
+
+                // => "Formation of Significant Shell Pair List ML" <= //
+
+                // Significant ket shell pairs RS for bra shell pair PQ
+                // represents the merge of ML_P and ML_Q as defined in Oschenfeld
+                // Unordered set structure allows for automatic merging as new elements are added
+                
+                std::unordered_set<int> ML_PQ;
```

Comment:
See my comment on the stripeout variables. I have the same comment regarding the use of std::unordered_set<int> here as well.

Reply:
Ah, std::unordered_set is necessary here, since we may have repeats of significant RS shells when we loop over P and when we loop over Q. Using an unordered set allows O(1) time checking to see if the element is already there, as well as O(1) adding of any new elements. 

(See comment above the variable "Unordered set structure allows for automatic merging as new elements are added")